numpy/array_api/_array_object.py

```diff
@@ -292,100 +290,122 @@ def _validate_index(key, shape):
-                    or step < 0 and not (-size - 1 <= key.stop <= max(0, size - 1))):
-                    raise IndexError("Slices with out-of-bounds stop are not allowed in the array API namespace")
-            return key
-
-        elif isinstance(key, tuple):
-            key = tuple(Array._validate_index(idx, None) for idx in key)
-
-            for idx in key:
-                if (
-                    isinstance(idx, np.ndarray)
-                    and idx.dtype in _boolean_dtypes
-                    or isinstance(idx, (bool, np.bool_))
-                ):
-                    if len(key) == 1:
-                        return key
-                    raise IndexError(
-                        "Boolean array indices combined with other indices are not allowed in the array API namespace"
-                    )
-                if isinstance(idx, tuple):
-                    raise IndexError(
-                        "Nested tuple indices are not allowed in the array API namespace"
-                    )
-
-            if shape is None:
-                return key
-            n_ellipsis = key.count(...)
-            if n_ellipsis > 1:
-                return key
-            ellipsis_i = key.index(...) if n_ellipsis else len(key)
 
-            for idx, size in list(zip(key[:ellipsis_i], shape)) + list(
-                zip(key[:ellipsis_i:-1], shape[:ellipsis_i:-1])
-            ):
-                Array._validate_index(idx, (size,))
-            if n_ellipsis == 0 and len(key) < len(shape):
+        nonexpanding_key = []
+        single_axes = []
+        n_ellipsis = 0
+        key_has_mask = False
+        for i in _key:
+            if i is not None:
+                nonexpanding_key.append(i)
+                if isinstance(i, Array) or isinstance(i, np.ndarray):
+                    if i.dtype in _boolean_dtypes:
+                        key_has_mask = True
+                    single_axes.append(i)
+                else:
+                    # i must not be an array here, to avoid elementwise equals
+                    if i == Ellipsis:
+                        n_ellipsis += 1
+                    else:
+                        single_axes.append(i)
+
+        n_single_axes = len(single_axes)
+        if n_ellipsis > 1:
+            raise IndexError(
+                f"{key=} contains {n_ellipsis} ellipses (...), but should "
+                "contain no more than 1."
+            )
+        elif n_ellipsis != 1:
+            # Note boolean masks must be the sole index, which we check for
+            # later on.
+            if not key_has_mask and n_single_axes < self.ndim:
                 raise IndexError(
-                    "Multidimensional arrays must include an index for every axis or use an ellipsis"
+                    f"{key=} contains {n_single_axes} single-axis indices, "
+                    f"but should contain {self.ndim=} single-axis indices - "
+                    "an array key must index every axis."
```

Comment:
```suggestion
                    "an array index must index every axis."
```

I wouldn't use the word "key" to refer to an index.

Reply:
Yeah it's not ideal, I think I was avoiding saying index twice in two different ways. How about
```suggestion
                    "an array index must cover every axis."
```

Reply:
That's probably fine, although I think I prefer my original wording.

Reply:
I like mentioning ellipsis, because it offers the correct solution.  Maybe something like:
> The multidimensional index only specifies N dimensions, but the array has K dimensions.  If this is intentional, add an Ellipsis (`...`) which expands into as many slices (`:`) as necessary.

Saying how many dimensions were not indexed seems cute, but before we bike-shed a lot, I also think the original was just fine.

Reply:
Thanks yeah that makes sense @seberg, pushed an inspired change if you and @asmeurer would like to take a look.